### ISSUE TITLE
Issue building raspicam2_node on Ubuntu Mate 18.04

### ISSUE THREAD

**Issue by ryanewel** [christianrauch/raspicam2_node#3]
OS: Ubuntu Mate 18.04
Platform: Raspberry Pi 3B+
ROS Distro: ROS2 Dashing installed from https://github.com/ros2/ros2/releases/download/release-dashing-20190806/ros2-dashing-20190806-linux-bionic-armhf.tar.bz2

I installed ROS2 with the given bundle. Then took the following steps
```
mkdir -p ~/raspicam2_ws/src
cd ~/raspicam2_ws/src
git clone https://github.com/christianrauch/raspicam2_node.git
cd ~/raspicam2_ws
colcon build
```

The build fails with the following error:
```
Starting >>> raspicam2
--- stderr: raspicam2                             
/usr/bin/ld: /opt/vc/lib/libvchostif.a(vc_vchi_gencmd.c.o): undefined reference to symbol 'vcos_pthreads_map_errno'
/opt/vc/lib/libvcos.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [raspicam2_node] Error 1
make[1]: *** [CMakeFiles/raspicam2_node.dir/all] Error 2
make: *** [all] Error 2
---
Failed   <<< raspicam2	[ Exited with code 2 ]

Summary: 0 packages finished [2.54s]
  1 package failed: raspicam2
  1 package had stderr output: raspicam2
```

**Comment by christianrauch** [comment]
I've only tested the build on Raspbian.
It might be that you need to install additional libraries on Ubuntu. Have you tried to install https://github.com/raspberrypi/userland?

**Comment by ryanewel** [comment]
I tried the following
```
cd ~
git clone https://github.com/raspberrypi/userland.git
cd ~/userland
./buildme
cd ~/ros2_raspicam_ws
colcon build
```
And got the same error as before. I'll keep trying, maybe I can get the missing library from here: https://github.com/raspberrypi/firmware ?

**Comment by christianrauch** [comment]
The firmware repo contains the precompiled libraries and executables. They might not run on Ubuntu 18.04.

The userland repo contains the `raspistill` example (https://github.com/raspberrypi/userland/tree/master/host_applications/linux/apps/raspicam) that you can use to verify that the userland was compiled and installed properly. If the userland isn't properly installed, colcon will not be able to find the required libraries and headers.

How is the library at `/opt/vc/lib/libvchostif.a` installed on Ubuntu?

**Comment by ryanewel** [comment]
I think it was installed from the userland package. `raspistill` hangs when I run it so something is wrong with my raspicam set up.

**Comment by ryanewel** [comment]
Ok I got to go back and try again. Starting with a fresh Ubuntu Mate 18.04 server installation, I followed the instructions from here: https://wiki.ubuntu.com/ARM/RaspberryPi 
specifically:
```
sudo add-apt-repository ppa:ubuntu-raspi2/ppa
sudo apt-get update
sudo groupadd -f --system gpio
sudo usermod -a -G gpio $(whoami)
sudo groupadd -f --system i2c
sudo usermod -a -G i2c $(whoami)
sudo groupadd -f --system input
sudo usermod -a -G input $(whoami)
sudo groupadd -f --system spi
sudo usermod -a -G spi $(whoami)
sudo usermod -a -G video $(whoami)
sudo apt-get install libraspberrypi-bin libraspberrypi-dev
sudo ln -s /usr /opt/vc
```
This got the raspicam working, but when I tried to compile again I got the following error:
```
Starting >>> raspicam2
--- stderr: raspicam2                            
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c: In function ‘raspicamcontrol_set_annotate’:
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1597:4: error: unknown type name ‘MMAL_PARAMETER_CAMERA_ANNOTATE_V4_T’; did you mean ‘MMAL_PARAMETER_CAMERA_ANNOTATE_V3_T’?
    MMAL_PARAMETER_CAMERA_ANNOTATE_V4_T annotate =
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    MMAL_PARAMETER_CAMERA_ANNOTATE_V3_T
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1598:4: warning: braces around scalar initializer
    {{MMAL_PARAMETER_ANNOTATE, sizeof(MMAL_PARAMETER_CAMERA_ANNOTATE_V4_T)}};
    ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1598:4: note: (near initialization for ‘annotate’)
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1598:38: error: ‘MMAL_PARAMETER_CAMERA_ANNOTATE_V4_T’ undeclared (first use in this function); did you mean ‘MMAL_PARAMETER_CAMERA_ANNOTATE_V3_T’?
    {{MMAL_PARAMETER_ANNOTATE, sizeof(MMAL_PARAMETER_CAMERA_ANNOTATE_V4_T)}};
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                      MMAL_PARAMETER_CAMERA_ANNOTATE_V3_T
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1598:38: note: each undeclared identifier is reported only once for each function it appears in
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1598:31: warning: excess elements in scalar initializer
    {{MMAL_PARAMETER_ANNOTATE, sizeof(MMAL_PARAMETER_CAMERA_ANNOTATE_V4_T)}};
                               ^~~~~~
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1598:31: note: (near initialization for ‘annotate’)
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1604:16: error: ‘MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V4’ undeclared (first use in this function); did you mean ‘MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V2’?
       char tmp[MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V4];
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V2
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1607:15: error: request for member ‘enable’ in something not a structure or union
       annotate.enable = 1;
               ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1614:30: error: request for member ‘text’ in something not a structure or union
             strftime(annotate.text, MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3, string, &tm );
                              ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1619:29: error: request for member ‘text’ in something not a structure or union
             strncpy(annotate.text, string, MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3);
                             ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1621:18: error: request for member ‘text’ in something not a structure or union
          annotate.text[MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3-1] = '\0';
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1626:28: error: request for member ‘text’ in something not a structure or union
          if(strlen(annotate.text))
                            ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1634:26: error: request for member ‘text’ in something not a structure or union
          strncat(annotate.text, tmp, MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3 - strlen(annotate.text) - 1);
                          ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1634:92: error: request for member ‘text’ in something not a structure or union
          strncat(annotate.text, tmp, MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3 - strlen(annotate.text) - 1);
                                                                                            ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1639:28: error: request for member ‘text’ in something not a structure or union
          if(strlen(annotate.text))
                            ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1647:26: error: request for member ‘text’ in something not a structure or union
          strncat(annotate.text, tmp, MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3 - strlen(annotate.text) - 1);
                          ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1647:92: error: request for member ‘text’ in something not a structure or union
          strncat(annotate.text, tmp, MMAL_CAMERA_ANNOTATE_MAX_TEXT_LEN_V3 - strlen(annotate.text) - 1);
                                                                                            ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1651:18: error: request for member ‘show_shutter’ in something not a structure or union
          annotate.show_shutter = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1654:18: error: request for member ‘show_analog_gain’ in something not a structure or union
          annotate.show_analog_gain = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1657:18: error: request for member ‘show_lens’ in something not a structure or union
          annotate.show_lens = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1660:18: error: request for member ‘show_caf’ in something not a structure or union
          annotate.show_caf = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1663:18: error: request for member ‘show_motion’ in something not a structure or union
          annotate.show_motion = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1666:18: error: request for member ‘show_frame_num’ in something not a structure or union
          annotate.show_frame_num = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1669:18: error: request for member ‘enable_text_background’ in something not a structure or union
          annotate.enable_text_background = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1671:15: error: request for member ‘text_size’ in something not a structure or union
       annotate.text_size = text_size;
               ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1675:18: error: request for member ‘custom_text_colour’ in something not a structure or union
          annotate.custom_text_colour = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1676:18: error: request for member ‘custom_text_Y’ in something not a structure or union
          annotate.custom_text_Y = text_colour&0xff;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1677:18: error: request for member ‘custom_text_U’ in something not a structure or union
          annotate.custom_text_U = (text_colour>>8)&0xff;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1678:18: error: request for member ‘custom_text_V’ in something not a structure or union
          annotate.custom_text_V = (text_colour>>16)&0xff;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1681:18: error: request for member ‘custom_text_colour’ in something not a structure or union
          annotate.custom_text_colour = MMAL_FALSE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1685:18: error: request for member ‘custom_background_colour’ in something not a structure or union
          annotate.custom_background_colour = MMAL_TRUE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1686:18: error: request for member ‘custom_background_Y’ in something not a structure or union
          annotate.custom_background_Y = bg_colour&0xff;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1687:18: error: request for member ‘custom_background_U’ in something not a structure or union
          annotate.custom_background_U = (bg_colour>>8)&0xff;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1688:18: error: request for member ‘custom_background_V’ in something not a structure or union
          annotate.custom_background_V = (bg_colour>>16)&0xff;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1691:18: error: request for member ‘custom_background_colour’ in something not a structure or union
          annotate.custom_background_colour = MMAL_FALSE;
                  ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1693:15: error: request for member ‘justify’ in something not a structure or union
       annotate.justify = justify;
               ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1694:15: error: request for member ‘x_offset’ in something not a structure or union
       annotate.x_offset = x;
               ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1695:15: error: request for member ‘y_offset’ in something not a structure or union
       annotate.y_offset = y;
               ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1698:15: error: request for member ‘enable’ in something not a structure or union
       annotate.enable = 0;
               ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1700:80: error: request for member ‘hdr’ in something not a structure or union
    return mmal_status_to_int(mmal_port_parameter_set(camera->control, &annotate.hdr));
                                                                                ^
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c: In function ‘raspicamcontrol_set_gains’:
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1726:63: error: ‘MMAL_PARAMETER_ANALOG_GAIN’ undeclared (first use in this function); did you mean ‘MMAL_PARAMETER_LOGGING’?
    status = mmal_port_parameter_set_rational(camera->control, MMAL_PARAMETER_ANALOG_GAIN, rational);
                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~
                                                               MMAL_PARAMETER_LOGGING
/home/robo/raspicam_ws/src/raspicam2_node/src/RaspiCamControl.c:1731:63: error: ‘MMAL_PARAMETER_DIGITAL_GAIN’ undeclared (first use in this function); did you mean ‘MMAL_PARAMETER_ANALOG_GAIN’?
    status = mmal_port_parameter_set_rational(camera->control, MMAL_PARAMETER_DIGITAL_GAIN, rational);
                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                               MMAL_PARAMETER_ANALOG_GAIN
make[2]: *** [CMakeFiles/raspicam_common.dir/src/RaspiCamControl.c.o] Error 1
make[1]: *** [CMakeFiles/raspicam_common.dir/all] Error 2
make: *** [all] Error 2
---
Failed   <<< raspicam2  [ Exited with code 2 ]

Summary: 0 packages finished [1.67s]
  1 package failed: raspicam2
  1 package had stderr output: raspicam2
```
It looks like the video libraries in apt for ubuntu 18.04 are out of date so I tried to install from the userland repo
```
git clone https://github.com/raspberrypi/userland.git
cd userland/
./buildme
```
And when I tried to build the raspicam again I got the same error as before. Also commands like `raspistill` hang again.

**Comment by christianrauch** [comment]
With the same error, you mean the linking error `undefined reference to symbol 'vcos_pthreads_map_errno'`?

The compilation error you show seems to be caused from different API. This should be solved by using the version that is also used with Raspbian, i.e. this is the only version I am testing with.

About the linking error, I am not sure. It might be because a library is missing or you might have incompatible versions installed. Can you completely remove the libraries installed by Ubuntu and reinstall the one from the userland repo again?

Would it be able for you to test this on Raspbian to make sure, that the issue is not cause by something else?

**Comment by christianrauch** [comment]
I am trying to get the raspicam node running on Ubuntu myself, and getting the same error about: `/usr/bin/ld: /opt/vc/lib/libvchostif.a(vc_vchi_gencmd.c.o): undefined reference to symbol 'vcos_pthreads_map_errno'`.

I tried the VideoCore libraries from packages in `ppa:ubuntu-pi-flavour-makers/ppa` and via the [userland repo](https://github.com/raspberrypi/userland) with no avail.

@ryanewel Did you manage to compile the node on Ubuntu?

**Comment by christianrauch** [comment]
The issue has been be fixed by 476464d3dcbe958ff25bf53580eb42c54a89b222. See also the updated documentation on how to install the VideoCore libraries from ppa on Ubuntu.
